File: FlowCrypt/src/main/java/com/flowcrypt/email/service/actionqueue/actions/RegisterUserPublicKeyAction.java
/*
 * Business Source License 1.0 © 2017 FlowCrypt Limited (human@flowcrypt.com).
 * Use limitations apply. See https://github.com/FlowCrypt/flowcrypt-android/blob/master/LICENSE
 * Contributors: DenBond7
 */

package com.flowcrypt.email.service.actionqueue.actions;

import android.content.Context;
import android.os.Parcel;

import com.flowcrypt.email.api.retrofit.ApiHelper;
import com.flowcrypt.email.api.retrofit.ApiService;
import com.flowcrypt.email.api.retrofit.request.model.InitialLegacySubmitModel;
import com.flowcrypt.email.api.retrofit.response.attester.InitialLegacySubmitResponse;
import com.flowcrypt.email.util.exception.ApiException;

import java.io.IOException;

import retrofit2.Response;

/**
 * This action describes a task which registers a user public key
 * using API "https://attester.flowcrypt.com/initial/legacy_submit".
 *
 * @author Denis Bondarenko
 *         Date: 30.01.2018
 *         Time: 18:01
 *         E-mail: DenBond7@gmail.com
 */

public class RegisterUserPublicKeyAction extends Action {
    public static final Creator<RegisterUserPublicKeyAction> CREATOR = new Creator<RegisterUserPublicKeyAction>() {
        @Override
        public RegisterUserPublicKeyAction createFromParcel(Parcel source) {
            return new RegisterUserPublicKeyAction(source);
        }

        @Override
        public RegisterUserPublicKeyAction[] newArray(int size) {
            return new RegisterUserPublicKeyAction[size];
        }
    };

    private String publicKey;

    public RegisterUserPublicKeyAction(String email, String publicKey) {
        super(email, ActionType.REGISTER_USER_PUBLIC_KEY);
        this.publicKey = publicKey;
    }


    protected RegisterUserPublicKeyAction(Parcel in) {
        super(in);
        this.publicKey = in.readString();
    }

    @Override
    public void run(Context context) throws IOException {
        ApiService apiService = ApiHelper.getInstance(context).getRetrofit().create(ApiService.class);
        Response<InitialLegacySubmitResponse> response = apiService.postInitialLegacySubmit(
                new InitialLegacySubmitModel(email, publicKey)).execute();

        InitialLegacySubmitResponse initialLegacySubmitResponse = response.body();
        if (initialLegacySubmitResponse == null) {
            throw new IllegalArgumentException("The response is null!");
        }

        if (initialLegacySubmitResponse.getApiError() != null) {
            throw new ApiException(initialLegacySubmitResponse.getApiError());
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(this.publicKey);
    }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/service/actionqueue/actions/RegisterUserPublicKeyAction.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.service.actionqueue.actions

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.flowcrypt.email.api.retrofit.ApiHelper
import com.flowcrypt.email.api.retrofit.ApiService
import com.google.gson.annotations.SerializedName

/**
 * This action describes a task which registers a user public key
 * using API "https://flowcrypt.com/attester/initial/legacy_submit".
 *
 * @author Denis Bondarenko
 * Date: 30.01.2018
 * Time: 18:01
 * E-mail: DenBond7@gmail.com
 */
data class RegisterUserPublicKeyAction @JvmOverloads constructor(
  override var id: Long = 0,
  override var email: String? = null,
  override val version: Int = 0,
  private val publicKey: String
) : Action {
  @SerializedName(Action.TAG_NAME_ACTION_TYPE)
  override val type: Action.Type = Action.Type.REGISTER_USER_PUBLIC_KEY

  override suspend fun run(context: Context) {
    val apiService = ApiHelper.getInstance(context).retrofit.create(ApiService::class.java)
    /*val body = InitialLegacySubmitModel(email!!, publicKey)
    val response = apiService.postInitialLegacySubmit(body).execute()

    val (apiError) = response.body() ?: throw IllegalArgumentException("The response is null!")

    if (apiError != null) {
      val code = apiError.code ?: 0
      if (code < 400 || code >= 500) {
        throw ApiException(apiError)
      }
    }*/
  }

  constructor(source: Parcel) : this(
    source.readLong(),
    source.readString(),
    source.readInt(),
    source.readString()!!
  )

  override fun describeContents(): Int {
    return 0
  }

  override fun writeToParcel(dest: Parcel, flags: Int) =
    with(dest) {
      writeLong(id)
      writeString(email)
      writeInt(version)
      writeString(publicKey)
    }

  companion object {
    @JvmField
    val CREATOR: Parcelable.Creator<RegisterUserPublicKeyAction> =
      object : Parcelable.Creator<RegisterUserPublicKeyAction> {
        override fun createFromParcel(source: Parcel): RegisterUserPublicKeyAction =
          RegisterUserPublicKeyAction(source)

        override fun newArray(size: Int): Array<RegisterUserPublicKeyAction?> = arrayOfNulls(size)
      }
  }
}
